core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -187,40 +206,212 @@ public MeterBinder newMeterBinder(MeterIdPrefix idPrefix) {
 
     @Override
     public String toString() {
-        final StringBuilder buf = new StringBuilder();
-        buf.append("HealthCheckedEndpointGroup(all:[");
-        for (ServerConnection connection : allServers) {
-            buf.append(connection.endpoint).append(',');
-        }
-        buf.setCharAt(buf.length() - 1, ']');
-        buf.append(", healthy:[");
-        for (Endpoint endpoint : endpoints()) {
-            buf.append(endpoint).append(',');
-        }
-        buf.setCharAt(buf.length() - 1, ']');
-        buf.append(')');
-        return buf.toString();
+        return MoreObjects.toStringHelper(this)
+                          .add("chosen", endpoints())
+                          .add("candidates", delegate.endpoints())
+                          .toString();
     }
 
-    /**
-     * Returns whether an {@link Endpoint} is healthy or not.
-     */
-    @FunctionalInterface
-    public interface EndpointHealthChecker {
-        CompletableFuture<Boolean> isHealthy(Endpoint endpoint);
-    }
+    private final class DefaultHealthCheckerContext
+            extends AbstractExecutorService implements HealthCheckerContext, ScheduledExecutorService {
 
-    static final class ServerConnection {
+        private final Object lock = new Object();
         private final Endpoint endpoint;
-        private final EndpointHealthChecker healthChecker;
+        private final Map<Future<?>, Boolean> scheduledFutures = new IdentityHashMap<>();
+        @Nullable
+        private AsyncCloseable handle;
+        final CompletableFuture<?> initialCheckFuture = new CompletableFuture<>();
+        private boolean destroyed;
 
-        private ServerConnection(Endpoint endpoint, EndpointHealthChecker healthChecker) {
+        DefaultHealthCheckerContext(Endpoint endpoint) {
             this.endpoint = endpoint;
-            this.healthChecker = healthChecker;
         }
 
-        Endpoint endpoint() {
+        void init(AsyncCloseable handle) {
+            assert this.handle == null;
+            this.handle = handle;
+        }
+
+        CompletableFuture<?> destroy() {
+            assert handle != null : handle;
+            return handle.closeAsync().handle((unused1, unused2) -> {
+                synchronized (lock) {
+                    if (destroyed) {
+                        return null;
+                    }
+
+                    destroyed = true;
+
+                    // Cancel all scheduled tasks.
+                    // Make a copy to avoid a potential ConcurrentModificationException.
```

Comment:
We added a listener to the `ScheduledFuture`, so `.cancel()` here might trigger self removal in the loop, so we need to make a copy.

Reply:
Ah - probably good to put that in the comment since it's hard to see

`Make a copy to prevent ConcurrentModificationException when the future's handler removes it from scheduledFutures as a result of the cancellation, which may happen on this thread.`

Reply:
Good idea. Added the comment.